### PR TITLE
fix(mempool): Stop ignoring some transaction broadcasts

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Sometimes Zebra will overwrite transactions in the mempool gossip watch channel, and never broadcast them to the network.

Fixing this is important for reliably sending newly created `lightwalletd` transactions. (For all other transactions, nodes can get them from elsewhere on the network.)

### Specifications

Nodes are not required to broadcast new mempool transactions, but they should:
https://developer.bitcoin.org/devguide/p2p_network.html#transaction-broadcasting

Zebra does not rely on peers broadcasting transactions, it also actively crawls memory pools.

### Complex Code or Requirements

This PR fixes a concurrency bug by repeatedly checking for new transactions from the mempool before broadcasting any.

It has a limit on the number of times it will wait for new changes, which ensures we eventually broadcast and don't infinitely loop.

## Solution

- Combine some transaction IDs that arrive close together rather than overwriting and ignoring them
- Limit the number of times we will wait for new transaction IDs

Related changes:
- Reduce verbosity of transaction broadcast logging

## Review

This is a low-priority bug fix.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

If we're still dropping some transaction IDs, we might want to change from a watch channel to a broadcast channel. But based on my logs that's unlikely, I'm only seeing a new transaction every few seconds.

### CODE REVIEW

**Comment by teor2345** [comment]
The Docker jobs are likely to fail with a known Google PGP key error, feel free to restart all the Docker jobs.

**Comment by teor2345** [comment]
Ah yeah there it is:
https://github.com/ZcashFoundation/zebra/actions/runs/4277535424/jobs/7446412852#step:9:622

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
This PR looks like it has a GitHub status reporting issue, the "Setup update-to-tip test" job has actually finished, but it's not showing up as finished.

That job isn't required, so mergify should still work fine.

File: zebrad/src/components/mempool/gossip.rs [hunk 1]
```diff
@@ -37,14 +37,35 @@ where
...
+        let mut combined_changes = 1;
+
+        // once we get new data in the channel, broadcast to peers,
+        // the mempool automatically combines some transactions that arrive close together
         receiver.changed().await?;
+        let mut txs = receiver.borrow().clone();
+        tokio::task::yield_now().await;
+
+        // also combine transactions that arrived shortly after this one
+        while receiver.has_changed()? && combined_changes < MAX_CHANGES_BEFORE_SEND {
```
Comment (oxarbitrage) [inline zebrad/src/components/mempool/gossip.rs:49]:
```suggestion
        while receiver.has_changed()? && combined_changes <= MAX_CHANGES_BEFORE_SEND {
```

Do we want to loop 10 times or 9 ?

File: zebrad/src/components/mempool/gossip.rs [hunk 2]
```diff
@@ -37,14 +37,35 @@ where
...
+            // Correctness: reset has_changed(), don't hold the lock
```
Comment (oxarbitrage) [inline zebrad/src/components/mempool/gossip.rs]: This comment is not totally clear to me. When we are resetting and what ?

**Review by oxarbitrage — DISMISSED** [review]
Looks pretty good but how do we know this is doing what we need ? Any chance to do a test ?

File: zebrad/src/components/mempool/gossip.rs (hunk 1 above)
Comment (teor2345) [reply zebrad/src/components/mempool/gossip.rs:49]: I don't think it matters either way. We have already checked once before the loop, so it's 10 combined changes at the end of the loop.

**Comment by teor2345** [comment]
> Looks pretty good but how do we know this is doing what we need ? Any chance to do a test ?

I have looked at the log, and I haven't seen any combined changes yet. So I think it will only happen under load.

I don't think we can write a test for this without using something like `loom` to get an unusual execution order:
https://docs.rs/loom/latest/loom/

(We need to run both tasks at the same time, and have an update happen in-between two of the times that the gossip task checks for changes. We can't do that in test code easily.)

File: zebrad/src/components/mempool/gossip.rs (hunk 2 above)
Comment (teor2345) [inline zebrad/src/components/mempool/gossip.rs]:
```suggestion
            // Correctness
            // - set the has_changed() flag to false using borrow_and_update()
            // - clone() so we don't hold the watch channel lock while modifying txs 
```

Comment (teor2345) [reply zebrad/src/components/mempool/gossip.rs]: I expanded it in commit 09912976983923ec59dcecb366c0979a98ffa981. The relevant API docs are:
https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.borrow_and_update
https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.has_changed

File: zebrad/src/components/mempool/gossip.rs [hunk 3]
```diff
@@ -37,14 +37,37 @@ where
...
+        // the mempool automatically combines some transactions that arrive close together
         receiver.changed().await?;
+        let mut txs = receiver.borrow().clone();
+        tokio::task::yield_now().await;
+
+        // also combine transactions that arrived shortly after this one
+        while receiver.has_changed()? && combined_changes < MAX_CHANGES_BEFORE_SEND {
+            // Correctness
+            // - set the has_changed() flag to false using borrow_and_update()
+            // - clone() so we don't hold the watch channel lock while modifying txs
```
Comment (teor2345) [inline zebrad/src/components/mempool/gossip.rs]:
```suggestion
            // - clone() so we don't hold the watch channel lock while modifying txs
```

**Review by oxarbitrage — APPROVED** [review]

**Comment by teor2345** [comment]
@mergifyio refresh

**Comment by teor2345** [comment]
Failed in the merge queue with:
> ERROR: (gcloud.compute.instances.create-with-container) Could not fetch resource:
> - Constraint constraints/compute.vmExternalIpAccess violated for project 143793276228. Add instance projects/zealous-zebra/zones/us-central1-a/instances/sync-past-checkpoint-6260-merge-9ce0bb0 to the constraint to use external IP with it.

https://github.com/ZcashFoundation/zebra/actions/runs/4309018317/jobs/7516676191#step:10:84

Gustavo is working on an infrastructure fix.

To re-queue the PR, comment `@mergify refresh` once he's fixed it.

**Comment by teor2345** [comment]
@mergifyio refresh

**Pull request merged**